components/script/dom/mutationobserver.rs

```diff
@@ -37,4 +72,159 @@ impl MutationObserver {
+        for (i, mo) in notifyList.iter().enumerate() {
+            let queue = mo.record_queue.clone();
+            *mo.record_queue.borrow_mut() = Vec::new();
+            // TODO: Step 5.3 Remove all transient registered observers whose observer is mo.
+            if !queue.borrow().is_empty() {
+                let callback: MutationCallback = *mo.callback.deref();
+                return callback.Call_(mo, *queue.borrow_mut().deref(), mo, ExceptionHandling::Report);
+            }
+        }
+        // Step 6 not needed as Servo doesn't implement anything related to slots yet.
+        Ok(())
+    }
+
+    //https://dom.spec.whatwg.org/#queueing-a-mutation-record
+    //Queuing a mutation record
+    pub fn queue_a_mutation_record(target: &Node, attr_type: Mutation) {
+        // Step 1
+        let mut interestedObservers: Vec<Root<MutationObserver>> = vec![];
+        let mut pairedStrings: Vec<DOMString> = vec![];
+        // Step 2
+        let mut nodes: Vec<Root<Node>> = vec![];
+        for ancestor in target.inclusive_ancestors() {
+            nodes.push(ancestor);
+        }
+        // Step 3
+        for node in &nodes {
+            for registered_observer in node.registered_mutation_observers().borrow().iter() {
+                match attr_type {
+                	// TODO check for Mutations other than Attribute
+                    Mutation::Attribute { name, namespace, oldValue, newValue } => {
+                    	let condition1: bool = node != &Root::from_ref(target) &&
+                            !registered_observer.subtree.get();
+                        let condition2: bool = registered_observer.attributes.get() == false;
+                        let condition3: bool = registered_observer.attribute_filter != DOMRefCell::new(vec![]) &&
+                            (!registered_observer.attribute_filter.borrow().contains(&name.unwrap()) || match Some(namespace) { Some(_) => true });
+                    	if !condition1 && !condition2 && !condition3 {
+                            // Step 3.1
+                            if !interestedObservers.contains(registered_observer) {
+                                interestedObservers.push(Root::from_ref(registered_observer));
+                            }
+                            // Step 3.2
+                            if registered_observer.attribute_old_value.get() == true {
+                                pairedStrings.push(DOMString::from(oldValue));
+                            }
+                        }
+                    }
+            	}
+            }
+        }
+        // Step 4
+        let mut i = 0;
+        for observer in &interestedObservers {
+            //Step 4.1
+            let record = &MutationRecord::new(DOMString::from("Attribute"), target);
+            //Step 4.2
+            match attr_type {
+            	Mutation::Attribute{ name, namespace, oldValue, newValue } => {
+                    if name != "" && namespace != "" {
```

Comment:
You don't need to check the contents of name and namespace; since they are present in the Attribute structure, we can use them. If you _did_ need to check them, `&*name != ""` probably would work.